component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -0,0 +1,222 @@
+        try {
+            boolean isValidAssertion = jwtValidator.isValidAssertion(getSignedJWT(clientAssertion));
+            if (StringUtils.isEmpty(oAuthClientAuthnContext.getClientId())) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Couldn't find client Id in context. Hence returning false.");
+                }
+                return false;
+            }
+            if (!isValidAssertion) {
+                return false;
+            }
+            return true;
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+    }
+
+    /**
+     * Returns whether the incoming request can be handled by the particular authenticator.
+     *
+     * @param httpServletRequest      http servelet request
+     * @param content                 map of request body params
+     * @param oAuthClientAuthnContext oAuthClientAuthnContext
+     * @return true if the incoming request can be handled.
+     */
+    @Override
+    public boolean canAuthenticate(HttpServletRequest httpServletRequest, Map<String, List> content,
+                                   OAuthClientAuthnContext oAuthClientAuthnContext) {
+
+        String clientAssertionType = null;
+        String clientAssertion = null;
+        List oauthJWTAssertionType = content.get(OAUTH_JWT_ASSERTION_TYPE);
+        List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION_TYPE);
+        if (oauthJWTAssertionType != null && oauthJWTAssertionType.get(0) != null) {
+            clientAssertionType = oauthJWTAssertionType.get(0).toString();
+        }
+        if (oauthJWTAssertion != null && oauthJWTAssertion.get(0) != null) {
+            clientAssertion = oauthJWTAssertion.get(0).toString();
+        }
+
+        return isValidJWTClientAssertionRequest(clientAssertionType, clientAssertion);
+    }
+
+    /**
+     * Retrievs the client ID which is extracted from the JWT.
+     *
+     * @param httpServletRequest
+     * @param content
+     * @param oAuthClientAuthnContext
+     * @return jwt 'sub' value as the client id
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public String getClientId(HttpServletRequest httpServletRequest, Map<String, List> content,
+                              OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        SignedJWT signedJWT = null;
+        try {
+            List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION);
+            if (oauthJWTAssertion != null && oauthJWTAssertion != null) {
+                signedJWT = getSignedJWT(oauthJWTAssertion.toString());
+            }
+            ReadOnlyJWTClaimsSet claimsSet = jwtValidator.getClaimSet(signedJWT);
+            String jwtSubject = jwtValidator.resolveSubject(claimsSet);
+            if (jwtValidator.validateSubject(jwtSubject)) {
+                oAuthClientAuthnContext.setClientId(jwtSubject);
```

Comment:
You don't need to set client ID to context here. It gives a wrong impression. It will be set by the service. Just returning client id is enough.  But if you need any other properties which you dirive here you may need to store for future reference. Let's avoid setting client id to context here.